mgr/dashboard: Group similar alerts

abhidesai6 opened the pull request:
fixes : https://tracker.ceph.com/issues/72788




<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-ar72788-group-similar-alertsm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)
</details>

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/api/prometheus.service.ts:
```diff
@@ -79,6 +80,10 @@ export class PrometheusService {
   }
 
+  getGroupedAlerts() {
+    return this.http.get<GroupAlertmanagerAlert[]>(`${this.baseURL}/alertgroup`);
```
I think we can just do the update in existing api getAlerts()

As that I see using `cluster_filter` and it might be breaking there. So we need to update existing one:

```
 getAlerts(clusterFilteredAlerts = false, params = {}): Observable<AlertmanagerAlert[]> {
    params['cluster_filter'] = clusterFilteredAlerts;
    return this.http.get<AlertmanagerAlert[]>(this.baseURL, { params });
  }
  ```
  
  as well update in `promethues.py` to accept cluster filter same as in /alert
  
  ```
  @RESTController.Collection(method='GET', path='/alertgroup')
    def get_alertgroup(self, **params):
        return self.alert_proxy('GET', '/alerts/groups', params)
```

cc @nizamial09 @aaSharma14 ^^^ correct me if I am wrong

nizamial09 replied:
yup, this needs the `cluster_filter` and I don't know if we need to do something in the prometheus for this. @cloudbehl

cloudbehl replied:
I think we need to test it out. @aaSharma14 should be able to help here.

nizamial09 commented:
> suggest to use now summary instead of description

I am all for using summary but I think the summary itself needs to be more clearer. Right now it is just an expansion of the title which is of no use and doesn't say much about the alert (probably why we went with description...). We need a cleanup task to revist the alerts and improve the summary with the help of documentation team

cloudbehl commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/prometheus/active-alert-list/active-alert-list.component.ts:
```diff
@@ -96,6 +92,25 @@ export class ActiveAlertListComponent extends PrometheusListHelper implements On
+      },
+      ...this.innerColumns.slice(1),
+      {
+        name: $localize`Count`,
```
These are alert instances that are created. 
```suggestion
        name: $localize`Instance`,
```

abhidesai6 replied:
<img width="4578" height="734" alt="image" src="https://github.com/user-attachments/assets/593ddef3-f48b-40bb-a5ec-1db0f3bfb96d" />

cloudbehl reviewed (approved):
Minor comment, This is really good! Great usability improvement

@aaSharma14 can you test the cluster filter and see nothing is breaking.

abhidesai6 commented:
Fixed test failure

aaSharma14 reviewed (approved):
Tested Locally, Works fine with a multi-cluster setup and cluster filter works as expected. Thanks @abhidesai6 for this PR, Looks great.

abhidesai6 commented:
Thanks @aaSharma14 !
Fixed the issue here. identifier for the children table was pushed incorrect. 
I have updated the PR.
<img width="4578" height="1596" alt="image" src="https://github.com/user-attachments/assets/4d940cef-f703-4893-98b3-998073109dd4" />

nizamial09 commented:
@abhidesai6  While I like the improvement, I don't generally like the way it is structured (table in table in table). I'd say we can remove the innermost table by all means.

We could structure the table in a way that we can get rid of the Details table and show the most useful information at the Individual Alerts list. What I propose is something like below.

`Instance | Description | Severity | Status | Started | Ends At`

* `Instance` col values can also have a hyperlink that points to the `generatorURL`
* We can also remove the Started and EndsAt col if we want and maybe add a popover to `Status` that says `Active (since 8 hours ago)` , `Resolved at Sep 10` or `Resolved (1 min ago)` 


I don't think the other metadata's like `fstype`, `job`, `fingerprint` etc are that useful and people might be okay with them missing in the table. I am skeptical about the `cluster` name too since this table is not shown in the context of multi-cluster. It'll only show the alert from this cluster only.

cc: @afreen23 @cloudbehl @aaSharma14

nizamial09 commented:
Do we need to show `Severity` on the inner table. The parent table already covers the Severity and that Severity kind of holds true for all the items in the group so we may not need to repeat that. But the `State` can change depending on the `instance` so that might be still valid. I'd probably get rid of the `Severity` too from the inner tables.

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/prometheus/active-alert-list/active-alert-list.component.ts:
```diff
@@ -96,6 +92,25 @@ export class ActiveAlertListComponent extends PrometheusListHelper implements On
+      },
+      ...this.innerColumns.slice(1),
+      {
+        name: $localize`Instance`,
```
Agreed. On the context of Alerts, we already have an item called `Instance` which shows the instance where alert is being raised on and that can be confused with this name. So I also vote for `Occurence`

afreen23 commented:
> @abhidesai6 While I like the improvement, I don't generally like the way it is structured (table in table in table). I'd say we can remove the innermost table by all means.
> 
> We could structure the table in a way that we can get rid of the Details table and show the most useful information at the Individual Alerts list. What I propose is something like below.
> 
> `Instance | Description | Severity | Status | Started | Ends At`
> 
> * `Instance` col values can also have a hyperlink that points to the `generatorURL`
> * (Optional) We can also remove the Started and EndsAt col if we want and maybe add a popover to `Status` that says `Active (since 8 hours ago)` , `Resolved at Sep 10` or `Resolved (1 min ago)`
> 
> I don't think the other metadata's like `fstype`, `job`, `fingerprint` etc are that useful and people might be okay with them missing in the table. I am skeptical about the `cluster` name too since this table is not shown in the context of multi-cluster. It'll only show the alert from this cluster only.
> 
> cc: @afreen23 @cloudbehl @aaSharma14

Yes we discussed this in upstream sync. 
This is a quick fix actually just  (not great) , and that we can do it in parts- UX improvements.
This page needs refactoring and re-design which will increase scope and what we later want to adopt I see something similar to what we did for notification area.
So once those PR merges we can reuse some of ideas from there :)

Thanks

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/dashboard-v3/dashboard/dashboard-v3.component.scss:
```diff
@@ -32,6 +33,15 @@
+
+  .card-text .date {
+    display: inline-block;
+    min-width: layout.rem(220px);
```
I meant spacings variable in prev comment like
```
 margin-left: layout.$spacing-03;
```
https://carbondesignsystem.com/elements/spacing/overview/

And seems like the largest is 220px. Are we sure we want to use this much ?

abhidesai6 replied:
I tried the margin  approach as well. But it mismatches the vertical alignment of `Total Occurrence`  hence using the min width for the `Active since` part.
yes, we need 220px, it is the overall width for `Active since tag`

afreen23 replied:
Okay, lets keep then we would need more refactoring and change in UX otherwise.
